File: backend/go/summarizer.go
package main

// Go backup implementation of Code Keeper's PR file summarization logic.
// This mirrors the behavior of the Next.js/TypeScript version so the
// backend can generate useful summaries even if the web layer is unavailable.

import (
	"fmt"
	"regexp"
	"strings"
)

type GitHubPRFile struct {
	Filename  string `json:"filename"`
	Status    string `json:"status"` // added | modified | removed | renamed
	Additions int    `json:"additions"`
	Deletions int    `json:"deletions"`
	Changes   int    `json:"changes"`
	Patch     string `json:"patch,omitempty"`
}

type FileSummary struct {
	Path    string `json:"path"`
	Status  string `json:"status"`
	Summary string `json:"summary"`
}

type fileChangeAnalysis struct {
	Description string
	Additions   []string
	Deletions   []string
}

func isPackageLockFile(path string) bool {
	matched, _ := regexp.MatchString(`(?i)package-lock\.json$`, path)
	return matched
}

func summarizePackageLockChanges(file GitHubPRFile) string {
	addedVersions := len(regexp.MustCompile(`(?m)^\+\s+"version":\s*"[^"]+"`).FindAllString(file.Patch, -1))
	removedVersions := len(regexp.MustCompile(`(?m)^-\s+"version":\s*"[^"]+"`).FindAllString(file.Patch, -1))
	bumped := minInt(addedVersions, removedVersions)
	addedOnly := maxInt(0, addedVersions-bumped)
	removedOnly := maxInt(0, removedVersions-bumped)

	statusLabel := statusLabelFor(file.Status)
	return fmt.Sprintf("**%s**: `%s` — package changes (added %d, updated %d, removed %d)",
		statusLabel, file.Filename, addedOnly, bumped, removedOnly)
}

func extractMeaningfulChanges(patch string) fileChangeAnalysis {
	if patch == "" {
		return fileChangeAnalysis{}
	}

	var additions, deletions []string

	for _, line := range strings.Split(patch, "\n") {
		if strings.HasPrefix(line, "+++") || strings.HasPrefix(line, "---") || strings.HasPrefix(line, "@@") {
			continue
		}

		if strings.HasPrefix(line, "+") && strings.TrimSpace(line) != "+" {
			content := strings.TrimSpace(line[1:])
			if content != "" && !strings.HasPrefix(content, "//") && !strings.HasPrefix(content, "#") && !strings.HasPrefix(content, "*") {
				additions = append(additions, content)
			}
		} else if strings.HasPrefix(line, "-") && strings.TrimSpace(line) != "-" {
			content := strings.TrimSpace(line[1:])
			if content != "" && !strings.HasPrefix(content, "//") && !strings.HasPrefix(content, "#") && !strings.HasPrefix(content, "*") {
				deletions = append(deletions, content)
			}
		}
	}

	desc := generateChangeDescription(additions, deletions, patch)
	return fileChangeAnalysis{
		Description: desc,
		Additions:   additions,
		Deletions:   deletions,
	}
}

func generateChangeDescription(additions, deletions []string, fullPatch string) string {
	newFunction := regexp.MustCompile(`^\s*(function|const|let|var)\s+(\w+)\s*[=(]`)
	newClass := regexp.MustCompile(`^\s*class\s+(\w+)`)
	newImport := regexp.MustCompile(`^\s*import\s+.*from`)
	newExport := regexp.MustCompile(`^\s*export\s+(function|const|class|default)`)
	newType := regexp.MustCompile(`^\s*(type|interface)\s+(\w+)`)
	configChange := regexp.MustCompile(`^\s*["']?\w+["']?\s*:`)
	testCase := regexp.MustCompile(`^\s*(it|test|describe)\s*\(`)

	for _, addition := range additions {
		if newFunction.MatchString(addition) {
			match := newFunction.FindStringSubmatch(addition)
			kind := "function"
			if match != nil && match[1] != "function" {
				kind = "method"
			}
			name := ""
			if match != nil {
				name = match[2]
			}
			return fmt.Sprintf("Added %s `%s`", kind, name)
		}
		if newClass.MatchString(addition) {
			match := newClass.FindStringSubmatch(addition)
			name := ""
			if match != nil {
				name = match[1]
			}
			return fmt.Sprintf("Added class `%s`", name)
		}
		if newImport.MatchString(addition) {
			match := regexp.MustCompile(`import\s+(.*)\s+from\s+['"](.*)['"]`).FindStringSubmatch(addition)
			if match != nil {
				return fmt.Sprintf("Imported %s from %s", match[1], match[2])
			}
			return "Added import"
		}
		if newExport.MatchString(addition) {
			return "Added exports"
		}
		if newType.MatchString(addition) {
			match := newType.FindStringSubmatch(addition)
			kind := "type"
			name := ""
			if match != nil {
				kind = match[1]
				name = match[2]
			}
			return fmt.Sprintf("Added %s `%s`", kind, name)
		}
		if testCase.MatchString(addition) {
			return "Added test cases"
		}
		if configChange.MatchString(addition) {
			return "Updated configuration"
		}
	}

	if len(additions) > 0 && len(deletions) > 0 {
		return "Modified implementation"
	}
	if len(additions) > 0 {
		snippet := strings.Join(additions[:minInt(2, len(additions))], " | ")
		return "Added code: " + snippet
	}
	if len(deletions) > 0 {
		return "Removed code"
	}
	if strings.Contains(fullPatch, "Binary file") {
		return "Updated binary content"
	}
	return ""
}

func getChangeMagnitude(additions, deletions int) string {
	total := additions + deletions
	if total == 0 {
		return ""
	}
	if total < 10 {
		return "minor change"
	}
	if total < 50 {
		return "moderate change"
	}
	if total < 200 {
		return "significant change"
	}
	return "major refactor"
}

func statusLabelFor(status string) string {
	switch status {
	case "added":
		return "New"
	case "modified":
		return "Updated"
	case "removed":
		return "Deleted"
	case "renamed":
		return "Renamed"
	default:
		return strings.ToUpper(status)
	}
}

func buildFileSummary(file GitHubPRFile) FileSummary {
	if isPackageLockFile(file.Filename) {
		return FileSummary{
			Path:    file.Filename,
			Status:  file.Status,
			Summary: summarizePackageLockChanges(file),
		}
	}

	analysis := extractMeaningfulChanges(file.Patch)
	statusLabel := statusLabelFor(file.Status)

	var parts []string
	parts = append(parts, fmt.Sprintf("**%s**: `%s`", statusLabel, file.Filename))

	if file.Status != "removed" {
		magnitude := getChangeMagnitude(file.Additions, file.Deletions)
		if magnitude != "" {
			parts = append(parts, fmt.Sprintf("(%s)", magnitude))
		}
	}

	desc := analysis.Description

	if strings.HasPrefix(desc, "Added code:") {
		snippet := strings.TrimPrefix(desc, "Added code: ")
		prefix := ""
		if file.Additions+file.Deletions > 50 {
			prefix = "Introduces new logic. Highlights: "
		}
		parts = append(parts, "— "+prefix+snippet)
	} else if desc != "" {
		parts = append(parts, "— "+desc)
	} else if len(analysis.Additions) > 0 {
		max := minInt(2, len(analysis.Additions))
		snippet := strings.Join(analysis.Additions[:max], " | ")
		parts = append(parts, "— "+snippet)
	} else if strings.Contains(file.Patch, "Binary file") {
		parts = append(parts, "— binary content updated")
	}

	return FileSummary{
		Path:    file.Filename,
		Status:  file.Status,
		Summary: strings.Join(parts, " "),
	}
}

func SummarizeFiles(files []GitHubPRFile) []FileSummary {
	result := make([]FileSummary, 0, len(files))
	for _, f := range files {
		result = append(result, buildFileSummary(f))
	}
	return result
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: backend/python/summarizer.py
"""
Python backup implementation of Code Keeper's PR file summarization logic.

Given a list of GitHub PR file objects (shape compatible with the REST API),
this module produces short, human-readable summaries similar to the Next.js
implementation.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Literal, Dict, Any
import re


Status = Literal["added", "modified", "removed", "renamed"]


@dataclass
class GitHubPRFile:
    filename: str
    status: Status
    additions: int
    deletions: int
    changes: int
    patch: str | None = None


@dataclass
class FileSummary:
    path: str
    status: Status
    summary: str


def is_package_lock_file(path: str) -> bool:
    return bool(re.search(r"package-lock\.json$", path, re.IGNORECASE))


def summarize_package_lock_changes(file: GitHubPRFile) -> str:
    patch = file.patch or ""
    added_versions = len(re.findall(r'^\+\s+"version":\s*"[^"]+"', patch, re.MULTILINE))
    removed_versions = len(re.findall(r'^-\s+"version":\s*"[^"]+"', patch, re.MULTILINE))
    bumped = min(added_versions, removed_versions)
    added_only = max(0, added_versions - bumped)
    removed_only = max(0, removed_versions - bumped)

    if file.status == "added":
        label = "New"
    elif file.status == "modified":
        label = "Updated"
    elif file.status == "removed":
        label = "Deleted"
    else:
        label = "Renamed"

    return (
        f"**{label}**: `{file.filename}` — package changes "
        f"(added {added_only}, updated {bumped}, removed {removed_only})"
    )


def extract_meaningful_changes(patch: str | None) -> Dict[str, Any]:
    if not patch:
        return {"description": "", "additions": [], "deletions": []}

    additions: List[str] = []
    deletions: List[str] = []

    for line in patch.splitlines():
        if line.startswith(("+++", "---", "@@")):
            continue

        if line.startswith("+") and line.strip() != "+":
            content = line[1:].strip()
            if content and not content.startswith(("//", "#", "*")):
                additions.append(content)
        elif line.startswith("-") and line.strip() != "-":
            content = line[1:].strip()
            if content and not content.startswith(("//", "#", "*")):
                deletions.append(content)

    description = generate_change_description(additions, deletions, patch)
    return {"description": description, "additions": additions, "deletions": deletions}


def generate_change_description(
    additions: List[str], deletions: List[str], full_patch: str
) -> str:
    patterns = {
        "newFunction": re.compile(r"^\s*(function|const|let|var)\s+(\w+)\s*[=(]"),
        "newClass": re.compile(r"^\s*class\s+(\w+)"),
        "newImport": re.compile(r"^\s*import\s+.*from"),
        "newExport": re.compile(r"^\s*export\s+(function|const|class|default)"),
        "newType": re.compile(r"^\s*(type|interface)\s+(\w+)"),
        "configChange": re.compile(r'^\s*["\']?\w+["\']?\s*:'),
        "testCase": re.compile(r"^\s*(it|test|describe)\s*\("),
    }

    for addition in additions[:3]:
        if patterns["newFunction"].match(addition):
            match = patterns["newFunction"].match(addition)
            kind = "function" if match and match.group(1) == "function" else "method"
            name = match.group(2) if match else ""
            return f"Added {kind} `{name}`"

        if patterns["newClass"].match(addition):
            match = patterns["newClass"].match(addition)
            name = match.group(1) if match else ""
            return f"Added class `{name}`"

        if patterns["newImport"].match(addition):
            match = re.match(r"import\s+(.*)\s+from\s+['\"](.*)['\"]", addition)
            if match:
                return f"Imported {match.group(1)} from {match.group(2)}"
            return "Added import"

        if patterns["newExport"].match(addition):
            return "Added exports"

        if patterns["newType"].match(addition):
            match = patterns["newType"].match(addition)
            kind = match.group(1) if match else "type"
            name = match.group(2) if match else ""
            return f"Added {kind} `{name}`"

        if patterns["testCase"].match(addition):
            return "Added test cases"

        if patterns["configChange"].match(addition):
            return "Updated configuration"

    if additions and deletions:
        return "Modified implementation"
    if additions:
        snippet = " | ".join(additions[:2])
        return f"Added code: {snippet}"
    if deletions:
        return "Removed code"
    if "Binary file" in full_patch:
        return "Updated binary content"
    return ""


def get_change_magnitude(additions: int, deletions: int) -> str:
    total = additions + deletions
    if total == 0:
        return ""
    if total < 10:
        return "minor change"
    if total < 50:
        return "moderate change"
    if total < 200:
        return "significant change"
    return "major refactor"


def build_file_summary(file: GitHubPRFile) -> FileSummary:
    if is_package_lock_file(file.filename):
        return FileSummary(
            path=file.filename,
            status=file.status,
            summary=summarize_package_lock_changes(file),
        )

    analysis = extract_meaningful_changes(file.patch)
    parts: List[str] = []

    if file.status == "added":
        status_label = "New"
    elif file.status == "modified":
        status_label = "Updated"
    elif file.status == "removed":
        status_label = "Deleted"
    else:
        status_label = "Renamed"

    parts.append(f"**{status_label}**: `{file.filename}`")

    if file.status != "removed":
        magnitude = get_change_magnitude(file.additions, file.deletions)
        if magnitude:
            parts.append(f"({magnitude})")

    desc: str = analysis["description"]
    adds: List[str] = analysis["additions"]

    if desc.startswith("Added code:"):
        snippet = desc.replace("Added code: ", "")
        prefix = "Introduces new logic. Highlights: " if file.additions + file.deletions > 50 else ""
        parts.append(f"— {prefix}{snippet}")
    elif desc:
        parts.append(f"— {desc}")
    elif adds:
        snippet = adds[0] if len(adds) == 1 else " | ".join(adds[:2])
        parts.append(f"— {snippet}")
    elif file.patch and "Binary file" in file.patch:
        parts.append("— binary content updated")

    return FileSummary(path=file.filename, status=file.status, summary=" ".join(parts))


def summarize_files(raw_files: List[Dict[str, Any]]) -> List[FileSummary]:
    """Convert raw GitHub file payloads into FileSummary objects."""
    files = [
        GitHubPRFile(
            filename=f["filename"],
            status=f["status"],
            additions=f.get("additions", 0),
            deletions=f.get("deletions", 0),
            changes=f.get("changes", 0),
            patch=f.get("patch"),
        )
        for f in raw_files
    ]
    return [build_file_summary(file) for file in files]


__all__ = ["GitHubPRFile", "FileSummary", "summarize_files"]


